statsmodels/multivariate/multivariate_ols.py

```diff
@@ -0,0 +1,517 @@
+
+    results.loc["Roy’s greatest root", 'Value'] = fn(eigv1.max())
+
+    r = v - (p - q + 1)/2
+    u = (p*q - 2) / 4
+    df1 = p * q
+    if p*p + q*q - 5 > 0:
+        t = np.sqrt((p*p*q*q - 4) / (p*p + q*q - 5))
+    else:
+        t = 1
+    df2 = r*t - 2*u
+    lmd = results.loc["Wilks’ lambda", 'Value']
+    lmd = np.power(lmd, 1 / t)
+    F = (1 - lmd) / lmd * df2 / df1
+    results.loc["Wilks’ lambda", 'Num DF'] = df1
+    results.loc["Wilks’ lambda", 'Den DF'] = df2
+    results.loc["Wilks’ lambda", 'F Value'] = F
+    pval = stats.f.sf(F, df1, df2)
+    results.loc["Wilks’ lambda", 'Pr > F'] = pval
+
+    V = results.loc["Pillai’s trace", 'Value']
+    df1 = s * (2*m + s + 1)
+    df2 = s * (2*n + s + 1)
+    F = df2 / df1 * V / (s - V)
+    results.loc["Pillai’s trace", 'Num DF'] = df1
+    results.loc["Pillai’s trace", 'Den DF'] = df2
+    results.loc["Pillai’s trace", 'F Value'] = F
+    pval = stats.f.sf(F, df1, df2)
+    results.loc["Pillai’s trace", 'Pr > F'] = pval
+
+    U = results.loc["Hotelling-Lawley trace", 'Value']
+    if n > 0:
+        b = (p + 2*n) * (q + 2*n) / 2 / (2*n + 1) / (n - 1)
+        df1 = p * q
+        df2 = 4 + (p*q + 2) / (b - 1)
+        c = (df2 - 2) / 2 / n
+        F = df2 / df1 * U / c
+    else:
+        df1 = s * (2*m + s + 1)
+        df2 = s * (s*n + 1)
+        F = df2 / df1 / s * U
+    results.loc["Hotelling-Lawley trace", 'Num DF'] = df1
+    results.loc["Hotelling-Lawley trace", 'Den DF'] = df2
+    results.loc["Hotelling-Lawley trace", 'F Value'] = F
+    pval = stats.f.sf(F, df1, df2)
+    results.loc["Hotelling-Lawley trace", 'Pr > F'] = pval
+
+    sigma = results.loc["Roy’s greatest root", 'Value']
+    r = np.max([p, q])
+    df1 = r
+    df2 = v - r + q
+    F = df2 / df1 * sigma
+    results.loc["Roy’s greatest root", 'Num DF'] = df1
+    results.loc["Roy’s greatest root", 'Den DF'] = df2
+    results.loc["Roy’s greatest root", 'F Value'] = F
+    pval = stats.f.sf(F, df1, df2)
+    results.loc["Roy’s greatest root", 'Pr > F'] = pval
+    return results.iloc[:, [4, 2, 0, 1, 3]]
+
+
+def _multivariate_ols_test(hypotheses, fit_results, exog_names,
+                            endog_names):
+    def fn(L, M, C):
+        # .. [1] https://support.sas.com/documentation/cdl/en/statug/63033/HTML/default/viewer.htm#statug_introreg_sect012.htm
+        params, df_resid, inv_cov, sscpr = fit_results
+        # t1 = (L * params)M
+        t1 = L.dot(params).dot(M) - C
+        # H = t1'L(X'X)^L't1
+        t2 = L.dot(inv_cov).dot(L.T)
+        q = matrix_rank(t2)
+        H = t1.T.dot(inv(t2)).dot(t1)
+
+        # E = M'(Y'Y - B'(X'X)B)M
+        E = M.T.dot(sscpr).dot(M)
+        return E, H, q, df_resid
+
+    return _multivariate_test(hypotheses, exog_names, endog_names, fn)
+
+
+def _multivariate_test(hypotheses, exog_names, endog_names, fn):
+    k_xvar = len(exog_names)
+    k_yvar = len(endog_names)
+    results = {}
+    for i in hypotheses:
```

Comment:
in general I try to avoid one letter names if it goes over several lines, e.g.
`for hypo in hypotheses`
hypo is still short but more informative and doesn't sound too close, hypothesis and hypotheses, otherwise I often use singular/plural names.